newrelic-agent/src/test/java/com/newrelic/agent/service/module/TrackedAddSetSynchronizationTest.java

```diff
@@ -0,0 +1,156 @@
+        TrackedAddSet<String> target = new TrackedAddSet<>(1_000_000);
+        List<String> expectedValues = Collections.synchronizedList(new LinkedList<String>());
+        List<Set<String>> deltas = Collections.synchronizedList(new LinkedList<Set<String>>());
+        AtomicInteger value = new AtomicInteger(1234);
+        int resetSubmitCount = 0;
+        int acceptSubmitCount = 0;
+
+        // Stuff the executor queue. Note that the tasks will block on the latch.
+        // The latch is in place to ensure that a single thread can't complete its work
+        // before the next item is queued.
+        for (int i = 0; i < 500_000; i++) {
+            if (r.nextFloat() < 0.05) {
+                // capture the items added since the last reset.
+                resetSubmitCount += 1;
+                executorService.submit(new ResettingRunnable(latch, deltas, target));
+            } else {
+                // add a new item to the tracked set.
+                acceptSubmitCount += 1;
+                executorService.submit(new AddingRunnable(latch, value, expectedValues, target));
+            }
+        }
+
+        // release the initial tasks
+        latch.countDown();
+        // we're not queueing anything else, and we want to make sure everything is complete.
+        executorService.shutdown();
+        executorService.awaitTermination(10, TimeUnit.SECONDS);
+
+        // ensure a reset is the last thing we do.
+        resetSubmitCount += 1;
+        deltas.add(target.resetReturningAdded());
+
+        // conditions we test:
+        //  - the number of accept and reset operations matches what we expected
+        //  - the individual deltas are disjoint from each other
+        //  - the full set is a union of the individual deltas.
+
+        Set<String> fullSet = target.resetReturningAll();
+
+        assertEquals("test has an issue: not all runnables completed", resetSubmitCount, deltas.size());
+        assertEquals("test has an issue: not all runnables completed", acceptSubmitCount, expectedValues.size());
+        assertEquals("full set lost values somewhere", acceptSubmitCount, fullSet.size());
+
+        int sumOfDeltaSizes = 0;
+        Set<String> allDeltas = new HashSet<>();
+        for (Set<String> delta : deltas) {
+            sumOfDeltaSizes += delta.size();
+            allDeltas.addAll(delta);
+        }
+
+        // since `allDeltas` is a Set, any values in more than one delta will result in a smaller
+        // `allDeltas.size()` than `sumOfDeltaSizes`.
+        assertEquals("the same value was in different sets", allDeltas.size(), sumOfDeltaSizes);
+
+        StringBuilder builder = new StringBuilder("Missing expected values:");
+
+        for (String expected : expectedValues) {
+            if (!allDeltas.contains(expected)) {
+                builder.append(" ").append(expected);
+            }
+        }
+
+        assertEquals(builder.toString(), allDeltas, fullSet);
```

Comment:
This test goes deep!  The complexity of testing in this particular case I think is worth the trade-off.